File: components/cooptypes/src/contracts/marketplace/actions/onMktWoDecl.ts
import * as Permissions from '../../../common/permissions'
import type * as Marketplace from '../../../interfaces/marketplace'
import { Actors } from '../../../common'

/**
 * Callback от `soviet::cancelexprd` или decline-эффекта в soviet (Story 8.4).
 * Стирает запись проекта из RAM (терминал жизненного цикла); причина — в
 * журнале действий. Сигнатура совпадает с
 * DECLINE_CALLBACK_SIGNATURE из soviet.hpp. require_auth(_soviet).
 */
export const authorizations = [{ permissions: [Permissions.active], actor: Actors._contract }] as const

export const actionName = 'onmktwodecl'

/**
 * @interface
 */
export type IOnMktWoDecl = Marketplace.IOnMktWoDecl

File: components/contracts/cpp/marketplace/src/p.mkt.wroff/onmktwodecl.cpp
/**
 * @brief Callback от `soviet` после отказа в Протоколе совета о списании
 * скоропорта или истечения срока повестки (Story 8.4, p.mkt.wroff).
 *
 * Сигнатура `(coopname, hash, reason)` соответствует
 * `DECLINE_CALLBACK_SIGNATURE` (см. `cpp/lib/core/soviet/soviet.hpp:19`).
 * Контракт `soviet` вызывает action из `cancelexprd` (повестка просрочена)
 * или вручную через `decline*` actions от своего имени, поэтому единственно
 * допустимая авторизация — `_soviet`.
 *
 * Эффект:
 *  - Находит wroffprops по proposal_hash == `hash`, проверяет статус PROPOSED.
 *  - Терминал жизненного цикла: запись проекта стирается из RAM; причина
 *    отказа остаётся в журнале действий (аргумент reason) и решении совета.
 *
 * Без ledger2-движений.
 *
 * @ingroup public_marketplace_actions
 */
void marketplace::onmktwodecl(eosio::name coopname,
                               checksum256 hash,
                               std::string reason) {
  require_auth(_soviet);

  auto p = Marketplace::get_writeoff_proposal_by_hash_or_fail(coopname, hash);
  eosio::check(p.status == WroffStatus::PROPOSED,
               "Проект списания не находится на повестке (callback повторный или поздний)");

  Marketplace::erase_writeoff_proposal(coopname, p.id);
}
